File: shared/go_libs/validation/datetime.go
package validation

import (
	"strings"
	"time"
)

// Date validates that a value is or can be parsed as a date.
func Date(opts ...func(*dateOpts)) Validator {
	o := &dateOpts{
		layout: "2006-01-02", // ISO 8601
	}
	for _, opt := range opts {
		opt(o)
	}

	return ValidatorFunc(func(value any) ValidationResult {
		switch v := value.(type) {
		case time.Time:
			return Success(v)
		case string:
			s := strings.TrimSpace(v)
			if s == "" {
				return Failure(o.getErrorMessage())
			}

			t, err := time.Parse(o.layout, s)
			if err != nil {
				return Failure(o.getErrorMessage())
			}
			return Success(t)
		default:
			return Failure("value must be a string or time.Time")
		}
	})
}

type dateOpts struct {
	layout       string
	errorMessage string
}

func (o *dateOpts) getErrorMessage() string {
	if o.errorMessage != "" {
		return o.errorMessage
	}
	return "invalid date format, expected: " + o.layout
}

// WithDateLayout sets the expected date layout.
func WithDateLayout(layout string) func(*dateOpts) {
	return func(o *dateOpts) {
		o.layout = layout
	}
}

// WithDateError sets a custom error message for Date.
func WithDateError(msg string) func(*dateOpts) {
	return func(o *dateOpts) {
		o.errorMessage = msg
	}
}

// DateTime validates that a value is or can be parsed as a datetime.
func DateTime(opts ...func(*dateTimeOpts)) Validator {
	o := &dateTimeOpts{
		layout: "2006-01-02T15:04:05", // ISO 8601
	}
	for _, opt := range opts {
		opt(o)
	}

	return ValidatorFunc(func(value any) ValidationResult {
		switch v := value.(type) {
		case time.Time:
			return Success(v)
		case string:
			s := strings.TrimSpace(v)
			if s == "" {
				return Failure(o.getErrorMessage())
			}

			t, err := time.Parse(o.layout, s)
			if err != nil {
				return Failure(o.getErrorMessage())
			}
			return Success(t)
		default:
			return Failure("value must be a string or time.Time")
		}
	})
}

type dateTimeOpts struct {
	layout       string
	errorMessage string
}

func (o *dateTimeOpts) getErrorMessage() string {
	if o.errorMessage != "" {
		return o.errorMessage
	}
	return "invalid datetime format, expected: " + o.layout
}

// WithDateTimeLayout sets the expected datetime layout.
func WithDateTimeLayout(layout string) func(*dateTimeOpts) {
	return func(o *dateTimeOpts) {
		o.layout = layout
	}
}

// WithDateTimeError sets a custom error message for DateTime.
func WithDateTimeError(msg string) func(*dateTimeOpts) {
	return func(o *dateTimeOpts) {
		o.errorMessage = msg
	}
}

// Time validates that a value is or can be parsed as a time.
func Time(opts ...func(*timeOpts)) Validator {
	o := &timeOpts{
		layout: "15:04:05",
	}
	for _, opt := range opts {
		opt(o)
	}

	return ValidatorFunc(func(value any) ValidationResult {
		switch v := value.(type) {
		case time.Time:
			return Success(v)
		case string:
			s := strings.TrimSpace(v)
			if s == "" {
				return Failure(o.getErrorMessage())
			}

			// Parse as time only (use reference date)
			t, err := time.Parse(o.layout, s)
			if err != nil {
				return Failure(o.getErrorMessage())
			}
			return Success(t)
		default:
			return Failure("value must be a string or time.Time")
		}
	})
}

type timeOpts struct {
	layout       string
	errorMessage string
}

func (o *timeOpts) getErrorMessage() string {
	if o.errorMessage != "" {
		return o.errorMessage
	}
	return "invalid time format, expected: " + o.layout
}

// WithTimeLayout sets the expected time layout.
func WithTimeLayout(layout string) func(*timeOpts) {
	return func(o *timeOpts) {
		o.layout = layout
	}
}

// WithTimeError sets a custom error message for Time.
func WithTimeError(msg string) func(*timeOpts) {
	return func(o *timeOpts) {
		o.errorMessage = msg
	}
}

// DateInRange validates that a date is within a specified range.
func DateInRange(minDate, maxDate time.Time, opts ...func(*dateRangeOpts)) Validator {
	o := &dateRangeOpts{
		layout:     "2006-01-02",
		hasMinDate: !minDate.IsZero(),
		hasMaxDate: !maxDate.IsZero(),
	}
	for _, opt := range opts {
		opt(o)
	}

	return ValidatorFunc(func(value any) ValidationResult {
		// First parse the date
		result := Date(WithDateLayout(o.layout)).Validate(value)
		if !result.IsValid {
			return result
		}

		dateValue := result.Value.(time.Time)

		// Check range
		if o.hasMinDate && dateValue.Before(minDate) {
			if o.errorMessage != "" {
				return Failure(o.errorMessage)
			}
			return Failuref("date must be on or after %s", minDate.Format(o.layout))
		}

		if o.hasMaxDate && dateValue.After(maxDate) {
			if o.errorMessage != "" {
				return Failure(o.errorMessage)
			}
			return Failuref("date must be on or before %s", maxDate.Format(o.layout))
		}

		return Success(dateValue)
	})
}

type dateRangeOpts struct {
	layout       string
	hasMinDate   bool
	hasMaxDate   bool
	errorMessage string
}

// WithDateRangeLayout sets the date layout for DateInRange.
func WithDateRangeLayout(layout string) func(*dateRangeOpts) {
	return func(o *dateRangeOpts) {
		o.layout = layout
	}
}

// WithDateRangeError sets a custom error message for DateInRange.
func WithDateRangeError(msg string) func(*dateRangeOpts) {
	return func(o *dateRangeOpts) {
		o.errorMessage = msg
	}
}

// DateAfter validates that a date is after a minimum date.
func DateAfter(minDate time.Time, opts ...func(*dateRangeOpts)) Validator {
	return DateInRange(minDate, time.Time{}, opts...)
}

// DateBefore validates that a date is before a maximum date.
func DateBefore(maxDate time.Time, opts ...func(*dateRangeOpts)) Validator {
	return DateInRange(time.Time{}, maxDate, opts...)
}

// RFC3339 validates datetime in RFC3339 format.
func RFC3339(opts ...func(*dateTimeOpts)) Validator {
	return DateTime(append(opts, WithDateTimeLayout(time.RFC3339))...)
}

// RFC3339Nano validates datetime in RFC3339Nano format.
func RFC3339Nano(opts ...func(*dateTimeOpts)) Validator {
	return DateTime(append(opts, WithDateTimeLayout(time.RFC3339Nano))...)
}

File: shared/shared/py_libs/py_libs/validation/datetime.py
"""
DateTime validators - PyDAL-style validators for date/time inputs.

Provides:
- IsDate: Validates date strings
- IsDateTime: Validates datetime strings
- IsTime: Validates time strings
- IsDateInRange: Validates date is within range
"""

from __future__ import annotations

from datetime import date, datetime, time
from typing import Union

from py_libs.validation.base import ValidationResult, Validator

# Type for date/time inputs
DateInput = Union[str, date, datetime]
TimeInput = Union[str, time, datetime]
DateTimeInput = Union[str, datetime]


class IsDate(Validator[DateInput, date]):
    """
    Validates that a value is or can be parsed as a date.

    Args:
        format: Expected date format (strptime format string)
                Default: "%Y-%m-%d" (ISO 8601)

    Example:
        validator = IsDate()
        result = validator("2024-01-15")  # Valid
        result = validator("15/01/2024")  # Invalid (wrong format)

        validator = IsDate(format="%d/%m/%Y")
        result = validator("15/01/2024")  # Valid
    """

    def __init__(
        self, format: str = "%Y-%m-%d", error_message: str | None = None
    ) -> None:
        self.format = format
        self.error_message = error_message

    def validate(self, value: DateInput) -> ValidationResult[date]:
        if isinstance(value, datetime):
            return ValidationResult.success(value.date())

        if isinstance(value, date):
            return ValidationResult.success(value)

        if isinstance(value, str):
            value = value.strip()
            if not value:
                return ValidationResult.failure(self._get_error_message())

            try:
                parsed = datetime.strptime(value, self.format)
                return ValidationResult.success(parsed.date())
            except ValueError:
                return ValidationResult.failure(self._get_error_message())

        return ValidationResult.failure("Value must be a string or date")

    def _get_error_message(self) -> str:
        return self.error_message or f"Invalid date format. Expected: {self.format}"


class IsDateTime(Validator[DateTimeInput, datetime]):
    """
    Validates that a value is or can be parsed as a datetime.

    Args:
        format: Expected datetime format (strptime format string)
                Default: "%Y-%m-%dT%H:%M:%S" (ISO 8601)

    Example:
        validator = IsDateTime()
        result = validator("2024-01-15T14:30:00")  # Valid
        result = validator("2024-01-15")           # Invalid (missing time)

        validator = IsDateTime(format="%Y-%m-%d %H:%M")
        result = validator("2024-01-15 14:30")     # Valid
    """

    def __init__(
        self,
        format: str = "%Y-%m-%dT%H:%M:%S",
        error_message: str | None = None,
    ) -> None:
        self.format = format
        self.error_message = error_message

    def validate(self, value: DateTimeInput) -> ValidationResult[datetime]:
        if isinstance(value, datetime):
            return ValidationResult.success(value)

        if isinstance(value, str):
            value = value.strip()
            if not value:
                return ValidationResult.failure(self._get_error_message())

            try:
                parsed = datetime.strptime(value, self.format)
                return ValidationResult.success(parsed)
            except ValueError:
                return ValidationResult.failure(self._get_error_message())

        return ValidationResult.failure("Value must be a string or datetime")

    def _get_error_message(self) -> str:
        return self.error_message or f"Invalid datetime format. Expected: {self.format}"


class IsTime(Validator[TimeInput, time]):
    """
    Validates that a value is or can be parsed as a time.

    Args:
        format: Expected time format (strptime format string)
                Default: "%H:%M:%S"

    Example:
        validator = IsTime()
        result = validator("14:30:00")  # Valid
        result = validator("14:30")     # Invalid (missing seconds)

        validator = IsTime(format="%H:%M")
        result = validator("14:30")     # Valid
    """

    def __init__(
        self, format: str = "%H:%M:%S", error_message: str | None = None
    ) -> None:
        self.format = format
        self.error_message = error_message

    def validate(self, value: TimeInput) -> ValidationResult[time]:
        if isinstance(value, datetime):
            return ValidationResult.success(value.time())

        if isinstance(value, time):
            return ValidationResult.success(value)

        if isinstance(value, str):
            value = value.strip()
            if not value:
                return ValidationResult.failure(self._get_error_message())

            try:
                parsed = datetime.strptime(value, self.format)
                return ValidationResult.success(parsed.time())
            except ValueError:
                return ValidationResult.failure(self._get_error_message())

        return ValidationResult.failure("Value must be a string or time")

    def _get_error_message(self) -> str:
        return self.error_message or f"Invalid time format. Expected: {self.format}"


class IsDateInRange(Validator[DateInput, date]):
    """
    Validates that a date is within a specified range.

    Args:
        min_date: Minimum date (inclusive), or None for no minimum
        max_date: Maximum date (inclusive), or None for no maximum
        format: Date format for parsing strings

    Example:
        from datetime import date

        validator = IsDateInRange(
            min_date=date(2024, 1, 1),
            max_date=date(2024, 12, 31)
        )
        result = validator("2024-06-15")  # Valid
        result = validator("2023-12-31")  # Invalid (before min)
    """

    def __init__(
        self,
        min_date: date | None = None,
        max_date: date | None = None,
        format: str = "%Y-%m-%d",
        error_message: str | None = None,
    ) -> None:
        self.min_date = min_date
        self.max_date = max_date
        self.format = format
        self.error_message = error_message

    def validate(self, value: DateInput) -> ValidationResult[date]:
        # First parse the date
        date_validator = IsDate(format=self.format)
        result = date_validator.validate(value)
        if not result.is_valid:
            return ValidationResult.failure(result.error or "Invalid date")

        date_value = result.value
        assert date_value is not None  # Type narrowing

        # Check range
        if self.min_date is not None and date_value < self.min_date:
            msg = self.error_message or f"Date must be on or after {self.min_date}"
            return ValidationResult.failure(msg)

        if self.max_date is not None and date_value > self.max_date:
            msg = self.error_message or f"Date must be on or before {self.max_date}"
            return ValidationResult.failure(msg)

        return ValidationResult.success(date_value)
